### PR TITLE
[App] Application logs in CLI

### PR DESCRIPTION
**Pull request by adam-lightning**
## What does this PR do?

This PR adds CLI command `lightning show logs <app_name> [<components>]` that prints logs

### Does your PR introduce any breaking changes? If yes, please list them.

We've extended the `run_app_in_cloud` context manager to return `app_name` as well. This is to enable easier testing, but might also affect other existing tests from different repositories.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

**Comment by williamFalcon** [comment]
is “show” the best command word here? @adam-lightning. 
how will this generalize as the CLI evolves?

**Comment by adam-lightning** [comment]
@williamFalcon  The idea is that `show` would be for showing / displaying / printing all information about given resource(s).
Something like just saying "Show me logs of X", "show me something ...", "show me my application" etc.

Tho, for the future, I'd distinguish the "show" from "list" as one first one answers the question about the object's contents / details, and the other lists the objects we might want to interact with.

Or You mean something else by the CLI evolution ?

**Comment by williamFalcon** [comment]
that makes sense. i'm just thinking about what else "show" will apply to other than logs

File: src/lightning_app/cli/lightning_cli.py
Comment (manskx) [inline src/lightning_app/cli/lightning_cli.py:44]: this should be required? 

should we also allow using `app_id`?

File: src/lightning_app/cli/lightning_cli.py
Comment (manskx) [inline src/lightning_app/cli/lightning_cli.py:47]: Docstrings and help strings will be nice. Specially for components to indicate the `flow` can be used

File: src/lightning_app/cli/lightning_cli.py
Comment (manskx) [inline src/lightning_app/cli/lightning_cli.py:66]: nit: show list of apps like above `Please select one of available: [{', '.join(apps.keys())}]`

File: src/lightning_app/utilities/app_logs.py
Comment (manskx) [inline src/lightning_app/utilities/app_logs.py:64]: nit: show the exception msg also

File: src/lightning_app/utilities/app_logs.py
Comment (manskx) [inline src/lightning_app/utilities/app_logs.py:52]: ❤️

File: src/lightning_app/utilities/app_logs.py
Comment (manskx) [inline src/lightning_app/utilities/app_logs.py:44]: nit: could you add an example of how the log message json looks like ?

**Comment by manskx** [comment]
@adam-lightning We should add two things:
- An issue/ticket to create update the docs
- Update changelog

**Comment by adam-lightning** [comment]
> @adam-lightning We should add two things:
> 
> * An issue/ticket to create update the docs
> * Update changelog

- Just create a ticket or also implement it before merging this PR ?
- I haven't found changelog in the existing files. Should I create one?

**Comment by manskx** [comment]
> > @adam-lightning We should add two things:
> > 
> > * An issue/ticket to create update the docs
> > * Update changelog
> 
> @manskx
> 
> * Just create a ticket or also implement it before merging this PR ?
> * I haven't found changelog in the existing files. Should I create one?

Would be nice to update the docs in the same PR  :))) up to you

changelog is here: https://github.com/Lightning-AI/lightning/blob/35ec79ad1f75790767a6857637c9bbe052e47b58/src/lightning_app/CHANGELOG.md

File: src/lightning_app/cli/lightning_cli.py
Comment (adam-lightning) [reply src/lightning_app/cli/lightning_cli.py:44]: We can think about using `app_id` that's a good idea actually... 
It's `required=False` because later we do a check anyway, and for better UX we list all available applications.  Pushing check into the `click` validator would mess up the code ...

If we do not provide the `app_name` we get something like 
```
Error: You have not specified any LightningApp. Please select one of available: [brawny-cerf-2556, lightning/hackernews-app-0.0.2]
```

File: src/lightning_app/utilities/logs_socket_api.py
Comment (adam-lightning) [reply src/lightning_app/utilities/logs_socket_api.py:33]: We agreed (at some point) that from the `lightning` side we should use only `lightning-cloud` as API spec, and everything above and custom, should be directly in lightning.

File: src/lightning_app/utilities/app_logs.py
Comment (adam-lightning) [reply src/lightning_app/utilities/app_logs.py:64]: Surfacing exception message might just confuse the user imho.
My idea was that the exception that I encountered are mapped, and we surface a nice error to the user.
If we added the exception message we would "leak" the underlying logic which is never recommended.

File: src/lightning_app/utilities/app_logs.py
Comment (adam-lightning) [reply src/lightning_app/utilities/app_logs.py:44]: I think that it's implicitly stated in the `_LogEvent` class. Including example directly might be also considered implementation leakage, so not sure if we want to include that explicitly here ;) 
```
{'message': '', 'timestamp': '', 'labels': {'app': '', 'container': '', 'filename': '', 'job': '', 'namespace': '', 'node_name': '', 'pod': '', 'stream': ''}}
```

File: src/lightning_app/cli/lightning_cli.py
Comment (adam-lightning) [reply src/lightning_app/cli/lightning_cli.py]: 👍

**Comment by adam-lightning** [comment]
> > > @adam-lightning We should add two things:
> > > 
> > > * An issue/ticket to create update the docs
> > > * Update changelog
> > 
> > 
> > @manskx
> > 
> > * Just create a ticket or also implement it before merging this PR ?
> > * I haven't found changelog in the existing files. Should I create one?
> 
> Would be nice to update the docs in the same PR :))) up to you
> 
> changelog is here: https://github.com/Lightning-AI/lightning/blob/35ec79ad1f75790767a6857637c9bbe052e47b58/src/lightning_app/CHANGELOG.md

Updated the changelog. But I don't really know how / what to update in docs tbh. Do Yo have any idea @manskx  as to what to write and where? I think that adding CLI command and documenting it in CLI is good enough, right ?

**Review by manskx — APPROVED** [review]
Great work!

**Comment by manskx** [comment]
> > > > @adam-lightning We should add two things:
> > > > 
> > > > * An issue/ticket to create update the docs
> > > > * Update changelog
> > > 
> > > 
> > > @manskx
> > > 
> > > * Just create a ticket or also implement it before merging this PR ?
> > > * I haven't found changelog in the existing files. Should I create one?
> > 
> > 
> > Would be nice to update the docs in the same PR :))) up to you
> > changelog is here: https://github.com/Lightning-AI/lightning/blob/35ec79ad1f75790767a6857637c9bbe052e47b58/src/lightning_app/CHANGELOG.md
> 
> Updated the changelog. But I don't really know how / what to update in docs tbh. Do Yo have any idea @manskx as to what to write and where? I think that adding CLI command and documenting it in CLI is good enough, right ?


I think we can create a GH issue to update the docs of this functionality.
cc: @tchaton @Felonious-Spellfire

Let's also keep an eye on the e2e CI after this gets merged.

File: src/lightning_app/cli/lightning_cli.py
Comment (tchaton) [inline src/lightning_app/cli/lightning_cli.py:54]: Why this change? I should belong to another PR.

File: src/lightning_app/cli/lightning_cli.py
Comment (tchaton) [inline src/lightning_app/cli/lightning_cli.py:64]: This doesn't work locally. So we need a mechanism to inform the users.

File: src/lightning_app/cli/lightning_cli.py
Comment (tchaton) [inline src/lightning_app/cli/lightning_cli.py:69]:
```suggestion
    """Show cloud application logs. By default prints logs for all currently available components.
```

File: src/lightning_app/cli/lightning_cli.py
Comment (tchaton) [inline src/lightning_app/cli/lightning_cli.py:101]:
```suggestion
            f"You have not specified any Lightning App. Please select one of available: [{', '.join(apps.keys())}]"
```

File: src/lightning_app/cli/lightning_cli.py
Comment (tchaton) [inline src/lightning_app/cli/lightning_cli.py:97]:
```suggestion
        raise click.ClickException("You don't have any application in the cloud. Please, run an application first with `--cloud`.")
```

Comment (tchaton) [inline src/lightning_app/cli/lightning_cli.py:97]: Does this include stopped apps or only running ones?

File: src/lightning_app/cli/lightning_cli.py
Comment (tchaton) [inline src/lightning_app/cli/lightning_cli.py:106]:
```suggestion
            f"The Lightning App '{app_name}' does not exist. Please select one of following: [{', '.join(apps.keys())}]"
```

File: src/lightning_app/utilities/app_logs.py
Comment (tchaton) [inline src/lightning_app/utilities/app_logs.py:97]: If I understand this properly, this doesn't strictly ensure the logs are in the correct order right ?

**Review by tchaton — APPROVED** [review]
LGTM once the comments are addressed.

File: src/lightning_app/utilities/app_logs.py
Comment (adam-lightning) [reply src/lightning_app/utilities/app_logs.py:97]: Yes, not strictly, but works in practice good enough.

Because each component has logs from different websocket we need some way of funnelling them into one stream (stdout), idea here is to push from all WSs to one priority queue (with priority on timestamp) so it get's sorted "on-the-fly".

Not ideal - we can imagine that network latency might mess up output, but it's the best what we can do, while actually doing it "on-the-fly" rather than pulling all logs -> sorting -> printing - which would be time and memory consuming.
Also I think it might be slightly better than UI because of this priority queue, I sometimes observed that UI had messed up the order of logs.

File: src/lightning_app/cli/lightning_cli.py
Comment (adam-lightning) [reply src/lightning_app/cli/lightning_cli.py:54]: You are absolutely right. Mea culpa here 😓

File: src/lightning_app/cli/lightning_cli.py
Comment (adam-lightning) [reply src/lightning_app/cli/lightning_cli.py:64]: @tchaton  Actually, now I think about it. Maybe we could unify the behaviour of local and cloud app running? It would take some time but we have (almost) all peaces of the puzzle.
We could locally stream all logs to some output file (e.g. somewhere in `.lightning` directory) and read them with `show logs`.

So by default running locally would actually return control right after starting an app instead of hanging and printing output. 

```
lightning run app app.py [--cloud] 

lightning show logs --follow.  # print logs as they go -> we have no difference between behaviour locally and on cloud
```

What do you think? Anywho - something to think about :D

File: src/lightning_app/cli/lightning_cli.py
Comment (adam-lightning) [reply src/lightning_app/cli/lightning_cli.py:97]: non-deleted ones. So getting logs from stopped will be ok

File: src/lightning_app/utilities/app_logs.py
Comment (awaelchli) [inline src/lightning_app/utilities/app_logs.py:76]: This probably works well for some apps, and is a good start. I just want to point out that an app doesn't need to have a constant amount of "components". These could come and go during runtime of the app. So if you "follow" the logs, that's probably useful to know.

**Review by awaelchli — APPROVED** [review]
I like "show". For a different product, we had a CLI redesign planned (now on ice) where "show" was supposed to show information about many different types resources. It also fits well for logs. 

`lightning show [what] [name or id]`

File: src/lightning_app/cli/lightning_cli.py
Comment (tchaton) [reply src/lightning_app/cli/lightning_cli.py:64]: Yes, why not? Personally, I think it is a bit of overkill in the grand scheme of all the things to build, but I can imagine some users using this feature to isolate logs from a machine.

File: src/lightning_app/utilities/app_logs.py
Comment (tchaton) [reply src/lightning_app/utilities/app_logs.py:97]: I think we could have a moving window buffer acting to sort the logs with a period of several seconds.

File: src/lightning_app/cli/lightning_cli.py
Comment (tchaton) [reply src/lightning_app/cli/lightning_cli.py:78]: Done !

File: src/lightning_app/cli/lightning_cli.py
Comment (tchaton) [reply src/lightning_app/cli/lightning_cli.py:83]: Done !

File: src/lightning_app/cli/lightning_cli.py
Comment (tchaton) [reply src/lightning_app/cli/lightning_cli.py:85]: Done !

File: src/lightning_app/utilities/app_logs.py
Comment (tchaton) [reply src/lightning_app/utilities/app_logs.py:76]: Yes, this is definitely a skateboard version and needs to be improved upon.

**Pull request merged**
